Add MKGROUP option for XPENDING, XREADGROUP, XCLAIM and XAUTOCLAIM commands for the case if the group doesn't already exist.

hwware opened the pull request:
This PR solve the issue https://github.com/redis/redis/issues/10630 and https://stackoverflow.com/questions/72158877/is-it-possible-to-let-the-redis-auto-create-the-group-name-if-it-not-exists-in-r

a consumer group should be created if it doesn't already exist when a client executes XPENDING or XREADGROUP.

And after discussing more, we could add this option for XCLAIM and XAUTOCLAIM commands as well.

Thus, a new option MKGROUP is added to these 4 commands, new command format as below:

XREADGROUP GROUP group consumer [COUNT count] [BLOCK milliseconds] [NOACK] **[MKGROUP]** STREAMS key [key ...] id [id ...]
XPENDING key group [ [IDLE min-idle-time] start end count [consumer]] **[MKGROUP]**
XCLAIM key group consumer min-idle-time id [id ...] [IDLE ms] [TIME unix-time-milliseconds] [RETRYCOUNT count] [FORCE] [JUSTID]**[MKGROUP]**
XAUTOCLAIM key group consumer min-idle-time start [COUNT count] [JUSTID] **[MKGROUP]**

guybe7 commented:
Why would an application execute these commands on a non existing group? Sounds like a bug in the application

maor-rosenfeld commented:
> Why would an application execute these commands on a non existing group? Sounds like a bug in the application

I'm not familiar with redis guidelines on automatic creation of resources, but at least in streams we can see that things are created automatically. For example, you can use [XADD](https://redis.io/commands/xadd/) to add a message to a stream even if said stream doesn't exist. Is that a bug in the application? Of course not.

So a stream is automatically created on some of the stream related commands. A consumer is also created automatically on [XREADGROUP](https://redis.io/commands/xreadgroup/). I'd expect the same behavior to apply to consumer groups.

This allows the application to avoid having to write recovery code to deal with a cluster reset (i.e. recreate all the consumer group), and to not worry about creating the groups on startup if they don't exist.

oranagra commented:
the fact that XADD creates a key is consistent with the rest of redis (same as LPUSH, HSET, etc).
looking at all the stream commands, it seems that consumers are created implicitly and groups must be created explicitly.
i'm not sure i understand why maybe @itamarhaber or @guybe7 do.
in any case, maybe if we want to change that, in a backwards compatible way, we need to add an MKGOUP argument (like the MKSTREADM arg we have in XGROUP CREATE

hwware commented:
> the fact that XADD creates a key is consistent with the rest of redis (same as LPUSH, HSET, etc). looking at all the stream commands, it seems that consumers are created implicitly and groups must be created explicitly. i'm not sure i understand why maybe @itamarhaber or @guybe7 do. in any case, maybe if we want to change that, in a backwards compatible way, we need to add an MKGOUP argument (like the MKSTREADM arg we have in XGROUP CREATE

@maor-rosenfeld I agree with what Oran suggests, we could add an MKGOUP argument for the XREADGROUP.  Of course, we could add another standlone command to check if the group exists. How do you think? Thanks

guybe7 commented:
IMHO it makes sense for `XADD` and `XGROUP CREATE` to have the ability to create a stream because in their nature they are write-commands.
it feels a bit weird for a read-command such as `XPENDING` or `XREADGROUP` to create a group (also, where does it stop? why not `XCLAIM`, XAUTOCLAIM`, etc.?)

guybe7 commented:
having said that, if the fact that some stream "read" commands don't create consumer groups makes it difficult for users we can go ahead with the optional MKSTREAM, but let's do it for all groups-related commands and in the future, we will make sure to always create the group if needed (i.e. get Redis will not reply with "no such group" in all future commands)

oranagra commented:
ok. let's do that for 7.2

hwware commented:
> MKGROUP sounds good to me. Creating a group on read might look weird, but a consumer group is essentially a read entity by nature. So it's only natural that the groups will be created automatically on read commands.

I will update this PR for adding an extra optional parameter MKGROUP for XPENGDING and XREADGROUP commands, Thanks

guybe7 commented on src/commands/xpending.json:
```diff
@@ -81,6 +81,13 @@
                 ]
+            },
+	    {
+                "name": "nomkgroup",
+                "token": "NOMKGROUP",
+                "type": "pure-token",
+                "optional": true,
+		"since":"7.2.0"
```
you should use MKGROUP (we want the default behavior to be avoiding creating a group, to keep backward compatibility)

hwware replied:
Thanks, I will update it.

hwware commented:
@guybe7 @maor-rosenfeld @oranagra 
I already finsih the code part for the new command option, please take a look and later I will add test cases for them, Thanks

maor-rosenfeld commented:
> @guybe7 @maor-rosenfeld @oranagra I already finsih the code part for the new command option, please take a look and later I will add test cases for them, Thanks

I'm not familiar with redis code but it looks good to me. Do you think the same should happen with XCLAIM and XAUTOCLAIM?

hwware commented:
> XAUTOCLAIM

@guybe7 @maor-rosenfeld When I read the doc of command XAUTOCLAIM , I notice it says that Conceptually, XAUTOCLAIM is equivalent to calling [XPENDING] and then [XCLAIM].
Personally,  I think it makes sense adding similar function which create consumer group if it doesn't exists.

Beside this, I have another idea that: 
add a new command:  XEXIST key group -- which indicates if there is one consumer group for one specific key, return 1 if it exists and return 0 if not. 
Thus, it provides client side has another option to run the command: XPENDING, XREADGROUP, XCLAIM, XAUTOCLAIM without adding the MKGROUP option.  Thanks

How do you think?

hwware commented:
> > > XAUTOCLAIM
> > 
> > 
> > @guybe7 @maor-rosenfeld When I read the doc of command XAUTOCLAIM , I notice it says that Conceptually, XAUTOCLAIM is equivalent to calling [XPENDING] and then [XCLAIM]. Personally, I think it makes sense adding similar function which create consumer group if it doesn't exists.
> > Beside this, I have another idea that: add a new command: XEXIST key group -- which indicates if there is one consumer group for one specific key, return 1 if it exists and return 0 if not. Thus, it provides client side has another option to run the command: XPENDING, XREADGROUP, XCLAIM, XAUTOCLAIM without adding the MKGROUP option. Thanks
> > How do you think?
> 
> The problem with commands like XEXIST is that they're begging you to implement an unsafe operation in your code (XEXIST and then XGROUP CREATE). Such a command will be useless without a distributed lock which not everyone are using.
> 
> I agree that both XCLAIM and XAUTOCLAIM should create a group using the same MKGROUP flag.

I got your point, I will implement this parameter in XCLAIM and XAUTOCLAIM command, Thanks for your explanation!

hwware commented:
@oranagra @guybe7 All codes and test cases are finished, please take a look Thanks a lot.

oranagra commented:
@hwware thanks.
we'll have to revisit that when approaching 7.2 (can't make that change in 7.0)

hwware commented:
> @hwware thanks. we'll have to revisit that when approaching 7.2 (can't make that change in 7.0)

No problem, Thanks Oran

oranagra commented on src/commands/xreadgroup.json:
```diff
@@ -69,6 +75,13 @@
+                "type": "pure-token",
+                "optional": true,
+		"since":"7.2.0"
+            },
```
@hwware you have indentation issues, your editor inserts tabs in some places.
please go over all the modified files and fix them.

oranagra commented on src/t_stream.c:
```diff
@@ -1692,7 +1692,7 @@ size_t streamReplyWithRange(client *c, stream *s, streamID *start, streamID *end
     if (!(flags & STREAM_RWR_RAWENTRIES))
         arraylen_ptr = addReplyDeferredLen(c);
-    streamIteratorStart(&si,s,start,end,rev);
+    streamIteratorStart(&si,s,start,end,rev);    
```
added white space at the end of line that doesn't have any changes otherwise

oranagra commented on src/t_stream.c:
```diff
@@ -2222,6 +2223,13 @@ void xreadCommand(client *c) {
+        } else if (!strcasecmp(o,"MKGROUP")) {
+            if (!xreadgroup) {
+                addReplyError(c,"The MKGROUP option is only supported by "
+                                "XREADGROUP, XPENDING, XCLAIM, and XAUTOCLAIM. You called XREAD instead.");
```
i don't think we need to mention all the supporting commands, just mention is't not supported for XREAD

oranagra commented on src/t_stream.c:
```diff
@@ -2331,7 +2354,7 @@ void xreadCommand(client *c) {
                 /* We also want to serve a consumer in a consumer group
                  * synchronously in case the group top item delivered is smaller
-                 * than what the stream has inside. */
+                 * than what the stream has inside. */                
```
white space

oranagra commented on src/t_stream.c:
```diff
@@ -2375,7 +2398,7 @@ void xreadCommand(client *c) {
                                                         consumer->name);
                 }
-            }
+            }            
```
whitespace

oranagra commented on src/t_stream.c:
```diff
@@ -2385,8 +2408,8 @@ void xreadCommand(client *c) {
-
-     /* We replied synchronously! Set the top array len and return to caller. */
+    
+    /* We replied synchronously! Set the top array len and return to caller. */
```
whitespace

oranagra commented on src/t_stream.c:
```diff
@@ -2259,16 +2267,31 @@ void xreadCommand(client *c) {
-                                       "option",
-                                    (char*)key->ptr,(char*)groupname->ptr);
+            if (o == NULL) {
+                addReplyErrorFormat(c, "No such key '%s'",
```
let's keep the old error code / behavior (NOGROUP), with a detailed message saying the key is missing...

oranagra commented on src/t_stream.c:
```diff
@@ -2259,16 +2267,31 @@ void xreadCommand(client *c) {
+            if (o == NULL) {
+                addReplyErrorFormat(c, "No such key '%s'",
+                                     (char*)key->ptr);
                 goto cleanup;
```
indentation is wrong.

oranagra commented on src/t_stream.c:
```diff
@@ -2259,16 +2267,31 @@ void xreadCommand(client *c) {
+                                       "option",(char*)groupname->ptr);
+                  goto cleanup;
+	       } 
+	       streamID id;
```
indentation is wrong.

oranagra commented on src/t_stream.c:
```diff
@@ -2259,16 +2267,31 @@ void xreadCommand(client *c) {
+                 server.dirty++;
+                 notifyKeyspaceEvent(NOTIFY_STREAM,"xgroup-create",key,c->db->id);
+               } else {
+                 addReplyErrorFormat(c, "Create consumer group '%s' in XREADGROUP failed" , (char*)groupname->ptr);
```
why would it ever fail? we already know the group doesn't exist.
please drop this.

oranagra commented on src/t_stream.c:
```diff
@@ -2880,17 +2903,29 @@ void xpendingCommand(client *c) {
-     * syntax errors before any other error. */
-    if (c->argc >= 6) {
-        int startidx = 3; /* Without IDLE */
+    if(c->argc == 4) {
+	mkgroup = 1;
+    } else if (c->argc >= 6) {
+	int totalOption = c->argc-1;
+	int startidx = 3; /* Without IDLE */
```
indentation issues..

why do we set `mkgroup` when argc==4?
i don't like this assumption, it can break in the future.
i understand that the current command syntax forces us to count things in order to guess what's coming next, but at least let's actually match strings that we can.

oranagra commented on src/t_stream.c:
```diff
@@ -2903,6 +2938,7 @@ void xpendingCommand(client *c) {
             /* Search for rest of arguments after 'IDLE <idle>' */
             startidx += 2;
         }
+	totalOption -= startidx+2;
```
indentation

oranagra commented on src/t_stream.c:
```diff
@@ -2880,17 +2903,29 @@ void xpendingCommand(client *c) {
     int startex = 0, endex = 0;
+    int mkgroup = 0;   
 
+    robj *o = lookupKeyRead(c->db,key);
+    streamCG *group;
+
+    if (checkType(c,o,OBJ_STREAM)) return;
+    if (o == NULL) {
+        addReplyErrorFormat(c, "No such key '%s'",
+                                     (char*)key->ptr);
+        return;
+    }
```
i'd rather move the lookup (KSN) and type check to after processing the arguments and syntax errors.

oranagra commented on src/t_stream.c:
```diff
@@ -2923,24 +2959,35 @@ void xpendingCommand(client *c) {
+            server.dirty++;
+            notifyKeyspaceEvent(NOTIFY_STREAM,"xgroup-create",key,c->db->id);
+        } else {
+            addReplyErrorFormat(c, "Create consumer group '%s' in XPENDING failed" , (char*)groupname->ptr);
+            return;
```
again, error that can't actually happen. please drop (and fix indentation issues and tabs)

oranagra commented on src/t_stream.c:
```diff
@@ -3119,20 +3166,38 @@ void xclaimCommand(client *c) {
-                              (char*)c->argv[2]->ptr);
+    if (checkType(c,o,OBJ_STREAM)) return; /* Type error. */
+    if (o == NULL) {
+        addReplyErrorFormat(c, "No such key '%s'",
```
let's keep the old error code.

oranagra replied:
i'd suggest to move the type check and all creations to after argument parsing.
i know it wasn't the case, but let's improve.

oranagra commented on src/t_stream.c:
```diff
@@ -3119,20 +3166,38 @@ void xclaimCommand(client *c) {
+            server.dirty++;
+            notifyKeyspaceEvent(NOTIFY_STREAM,"xgroup-create",c->argv[1],c->db->id);
+        } else {
+            addReplyErrorFormat(c, "Create consumer group '%s' in CLAIMfailed" , (char*)c->argv[2]->ptr);
+	    return;
```
drop error that can't happen.

oranagra commented on src/t_stream.c:
```diff
@@ -3373,16 +3443,32 @@ void xautoclaimCommand(client *c) {
-                            (char*)c->argv[1]->ptr,
-                            (char*)c->argv[2]->ptr);
+    if(o == NULL) {
+        addReplyErrorFormat(c, "No such key '%s'",
```
let's keep the error code

oranagra commented on src/t_stream.c:
```diff
@@ -3373,16 +3443,32 @@ void xautoclaimCommand(client *c) {
+            server.dirty++;
+            notifyKeyspaceEvent(NOTIFY_STREAM,"xgroup-create",c->argv[1],c->db->id);
+        } else {
+            addReplyErrorFormat(c, "Create consumer group '%s' in CAUTOLAIM failed" , (char*)c->argv[2]->ptr);
+            return;
```
drop impossible error

hwware replied:
I will update the codes following your comments, and rebase it as well.  Thanks Oran

oranagra reviewed (commented):
sorry for the delay.. let's revive it for merge in time for 7.2.
i did one round of review (excluding tests), a closer look will be needed after these are addressed.

hwware commented on src/t_stream.c:
```diff
@@ -2880,68 +2899,90 @@ void xpendingCommand(client *c) {
+	  }
+       }
     }
 
```
Hi Oran, here I have one concern, but so far I have no better idea. Please comment it if you have better idea on it.
For the command xpending, it could be 
`XPENDING <key> <group> [<start> <stop> <count> [<consumer>]]`
or
`XPENDING <key> <group> [<start> <stop> <count> ] [MKGROUP]`

The last argument could be **consumer** or [MKGROUP], so the only way to distinguish is to use `strcasecmp(c->argv[j]->ptr, "MKGROUP")`.   If I move the MKGROUP augement to the front of the IDLE or `<start>`, it will break current change.
Thus, I choose my current code. 
If you have better arugment format or better coding logic, welcome to comment here.

oranagra replied:
i see that in 6.2 we chose to extend this command by adding optional arguments (the IDLE) before the `<start>`.
so in this case, since MKGROUP isn't related to the range specification, i think it should come right after `<group>`

hwware commented:
@oranagra Code is updated, please check them, Thanks

oranagra commented on src/commands/xpending.json:
```diff
@@ -81,6 +85,13 @@
+                "token": "MKGROUP",
+                "type": "pure-token",
+                "optional": true,
+                "since":"7.2.0"
```
as i noted here https://github.com/redis/redis/pull/10747#discussion_r1002706299
i think it should come before the `filters` block.

oranagra commented on src/t_stream.c:
```diff
@@ -2259,16 +2266,28 @@ void xreadCommand(client *c) {
                 goto cleanup;
-            }
-            groups[id_idx] = group;
+	    }
```
indentation

oranagra commented on src/t_stream.c:
```diff
@@ -2331,7 +2350,7 @@ void xreadCommand(client *c) {
                 /* We also want to serve a consumer in a consumer group
                  * synchronously in case the group top item delivered is smaller
-                 * than what the stream has inside. */
+                 * than what the stream has inside. */ 
```
white space

oranagra commented on src/t_stream.c:
```diff
@@ -2385,8 +2404,8 @@ void xreadCommand(client *c) {
     }
-
-     /* We replied synchronously! Set the top array len and return to caller. */
+          
```
white space

oranagra commented on src/t_stream.c:
```diff
@@ -2259,16 +2266,28 @@ void xreadCommand(client *c) {
+               streamID id;
+               id.ms = 0;
+               id.seq = 0;
+               group = streamCreateCG(o->ptr,groupname->ptr,strlen(groupname->ptr),&id,SCG_INVALID_ENTRIES_READ);
+               server.dirty++;
+               notifyKeyspaceEvent(NOTIFY_STREAM,"xgroup-create",key,c->db->id);
```
let's move the creation of the group to after argument parsing.
so that we know we're not making changes if the command is gonna fail on an error.

hwware replied:
Address this comment, create the group if necessary after the argyment parsing part, Please take a look

oranagra commented on src/t_stream.c:
```diff
@@ -2880,68 +2899,90 @@ void xpendingCommand(client *c) {
+    while(j < c->argc) {
+       if (!strcasecmp(c->argv[j]->ptr, "MKGROUP")) {
+          mkgroup = 1;
+	  j++;
+	  continue;
```
please make sure your code doesn't contain tabs.

oranagra commented on src/t_stream.c:
```diff
@@ -2880,68 +2899,90 @@ void xpendingCommand(client *c) {
+       }
+       if (!strcasecmp(c->argv[j]->ptr, "IDLE")) {
+          if (j+1 == c->argc) {
+            addReplyErrorObject(c,shared.syntaxerr);
```
indentation by 4, not 2.

hwware replied:
Check all code styles and update them.  Thanks a lot

oranagra reviewed (commented):
@hwware it's really hard for me to read the code and focus on the important things when there are so many white space issues, please fix them first.

other than that, i have some actual comments about the command syntax and side effects of failed calls, so please address that across the board, and i'll take another look later.

oranagra commented on src/t_stream.c:
```diff
@@ -2306,6 +2321,25 @@ void xreadCommand(client *c) {
+            int id_idx = i - streams_arg - streams_count;
+            robj *key = c->argv[i-streams_count];
+            robj *o = lookupKeyRead(c->db,key);
+            streamCG *group = streamLookupCG(o->ptr,groupname->ptr);
```
didn't we already lookup the key and group?
we saved the group into `groups`, there's no need to re-lookup.
let's do the same thing for the keys and also avoid the other lookupKeyRead below.

hwware replied:
Adressed this comment when I updated https://github.com/redis/redis/pull/10747#discussion_r1063563264

oranagra commented on src/t_stream.c:
```diff
@@ -2880,68 +2914,85 @@ void xpendingCommand(client *c) {
+            }
+            if (getLongLongFromObjectOrReply(c, c->argv[j+1], &minidle, NULL) == C_ERR)
                 return;
             if (c->argc < 8) {
```
isn't this outdated? let's match it to `j`

hwware replied:
adressed this comment when i updated following comment https://github.com/redis/redis/pull/10747#discussion_r1063563264

oranagra commented on src/t_stream.c:
```diff
@@ -2880,68 +2914,85 @@ void xpendingCommand(client *c) {
-            startidx += 2;
+            j = j+2;
+        }
+        if (j+2 >= c->argc) {
```
please add a comment ("expecting at least start, end, count to follow", right?)

hwware replied:
added the comment when I updated https://github.com/redis/redis/pull/10747#discussion_r1063563264

oranagra commented on src/t_stream.c:
```diff
@@ -3119,26 +3172,24 @@ void xclaimCommand(client *c) {
-    if (o) {
-        if (checkType(c,o,OBJ_STREAM)) return; /* Type error. */
-        group = streamLookupCG(o->ptr,c->argv[2]->ptr);
+    if(c->argc > 6 && !strcasecmp(c->argv[c->argc-1]->ptr,"MKGROUP")) {
+        mkgroup = 1;
```
optional arguments are interchangeable, you shouldn't search for it only at the last argument.
why not move it to the argument parsing? (where you already skip that arg anyway)

hwware replied:
updated this when I changes are made for https://github.com/redis/redis/pull/10747#discussion_r1063563264

oranagra commented on src/t_stream.c:
```diff
@@ -3119,26 +3172,24 @@ void xclaimCommand(client *c) {
         != C_OK) return;
     if (minidle < 0) minidle = 0;
 
+    if (o == NULL) {
+        addReplyErrorFormat(c, "-NOGROUP No such key '%s'",
+                                     (char*)c->argv[1]->ptr);
+        return;
+    } 
+    if (checkType(c,o,OBJ_STREAM)) return; /* Type error. */
```
let's move all of that (and the lookupKey at the top of this function) to after the argument parsing.
no sense in looking up the key if we have a syntax error.

hwware replied:
Hi Oran, I already update all code logic, the process will be first parse the argument, and then check the key, finally check the group. And also I fix all code styles, Please check the latest codes, Thanks for your hard working.

oranagra commented on tests/unit/type/stream-cgroups.tcl:
```diff
@@ -124,6 +124,20 @@ start_server {
         }
     }
 
+    test {XPENDING with MKGROUP option to create consumer group} {
+	r XADD mystream_extra * t 1
+	r XADD mystream_extra * s 2
+	catch {[r XPENDING mystream_extra mygroup2 - + 10]} err
+        assert_match {*No consumer group 'mygroup2'*} $err
+
+	set pending [r XPENDING mystream_extra mygroup2 MKGROUP - + 10]
+        assert {[llength $pending] == 0}
+	r XREADGROUP GROUP mygroup2 consumer-1 COUNT 10 STREAMS mystream_extra ">"
+	set pending [r XPENDING mystream_extra mygroup2 - + 10]
+        assert {[llength $pending] == 2}
```
the tests still has tabs in them, please change to spaces..

hwware replied:
Correcetd all intendation in the file.

oranagra commented on tests/unit/type/stream-cgroups.tcl:
```diff
@@ -509,6 +537,39 @@ start_server {
+        set id2 [r XADD mystream * b 2]
+        set id3 [r XADD mystream * c 3]
+
+        after 200
```
why do we need all these sleeps?

hwware replied:
I will address your comments in next a few days, sorry for that.

hwware replied:
updated @oranagra

hwware commented:
@oranagra Hi Oran, could you please take sometime to check my latest update?  Thanks a lot

hoyhoy commented:
We need this for testing our app.  Our unit tests rely on flushed database and without a MKGROUP, this causes them all to fail.  It would be good for production as well.  It's very odd indeed that streams are the only data type that have this behavior where you need to create a group before using the read commands -- SETs, ZSETs, LISTs, and HASHes don't require anything like this.

hwware commented:
Hi Oran,

I have updated the code and also addressed the comments and replied. And I have even updated json file for 8.0.
here are the commit details which had covered the comments.
https://github.com/redis/redis/pull/10747/commits/1693d83dc45245d49a0ac42fd621361d6e0d6e85
https://github.com/redis/redis/pull/10747/commits/5ca566454cbb87d749e3d23b4773642ab1c87429
https://github.com/redis/redis/pull/10747/commits/765129339487770b95823d8d9499eed9be0cef74
